pkg/flowkit/services/project.go

```diff
@@ -327,94 +319,93 @@ func (p *Project) Deploy(network string, update bool) ([]*contracts.Contract, er
 
 		sentTx, err := p.gateway.SendSignedTransaction(tx)
 		if err != nil {
 			p.logger.StopProgress()
-			p.logger.Error(fmt.Sprintf("%s error: %s", contract.Name(), err))
-			deployErr = true
+			deployErr.add(contract, err, "failed to send deployment transaction")
 			continue
 		}
 
 		result, err := p.gateway.GetTransactionResult(sentTx.ID(), true)
+		p.logger.StopProgress()
 		if err != nil {
-			p.logger.StopProgress()
-			p.logger.Error(fmt.Sprintf("%s error: %s", contract.Name(), err))
-			deployErr = true
+			deployErr.add(contract, err, "could not retrieve deployment result")
 			continue
 		}
-		if result == nil {
-			p.logger.Error("could not fetch the result of deployment, skipping")
-			deployErr = true
+		if result != nil && result.Error != nil {
+			deployErr.add(contract, result.Error, "failed deploying contract")
 			continue
 		}
-		if result.Error != nil {
-			deployErr = true
-			p.logger.StopProgress()
-			if exists && update {
-				p.logger.Error(fmt.Sprintf(
-					"Error updating %s: (%s)\n",
-					output.Red(contract.Name()),
-					result.Error.Error(),
-				))
-			} else {
-				p.logger.Error(fmt.Sprintf(
-					"Error deploying %s: (%s)\n",
-					output.Red(contract.Name()),
-					result.Error.Error(),
-				))
-			}
-		}
 
-		if result.Error == nil && !deployErr {
-			changeStatus := ""
-			if exists && update {
-				changeStatus = "(update)"
-			}
-			p.logger.StopProgress()
-			p.logger.Info(fmt.Sprintf(
-				"%s -> 0x%s (%s) %s\n",
-				output.Green(contract.Name()),
-				contract.Target(),
-				sentTx.ID().String(),
-				changeStatus,
-			))
+		changeStatus := ""
+		if exists && update {
+			changeStatus = "(updated)"
 		}
+
+		p.logger.Info(fmt.Sprintf(
+			"%s -> 0x%s (%s) %s\n",
+			output.Green(contract.Name()),
+			contract.Target(),
+			sentTx.ID().String(),
+			changeStatus,
+		))
 	}
 
-	if !deployErr {
-		if update && numOfUpdates > 0 {
-			p.logger.Info(fmt.Sprintf("%d contracts updated successfully", numOfUpdates))
-		}
-		p.logger.Info(fmt.Sprintf("\n%s All contracts deployed successfully", output.SuccessEmoji()))
-	} else {
-		err = fmt.Errorf("failed to deploy all contracts")
-		p.logger.Error(err.Error())
-		return nil, err
+	if len(deployErr.contracts) > 0 {
+		return nil, deployErr
 	}
 
+	p.logger.Info(fmt.Sprintf("\n%s All contracts deployed successfully", output.SuccessEmoji()))
 	return orderedContracts, nil
 }
+
+type ErrProjectDeploy struct {
+	contracts map[string]error
+}
+
+func (d *ErrProjectDeploy) add(contract *contracts.Contract, err error, msg string) {
+	if d.contracts == nil {
+		d.contracts = make(map[string]error)
+	}
+	d.contracts[contract.Name()] = fmt.Errorf("%s: %w", msg, err)
+}
+
+func (d *ErrProjectDeploy) Contracts() map[string]error {
+	return d.contracts
+}
+
+func (d *ErrProjectDeploy) Error() string {
+	err := ""
+	for c, e := range d.contracts {
+		err = fmt.Sprintf("%s %s: %s,", err, c, e.Error())
+	}
+	return err
+}
```

Comment:
I think that as long as it's used here it should be defined here, because you can easily read the code. If we start using it other places at that point we can extract it. Go is really pragmatic about this kind of stuff and I began to like that: don't predict the future too much, if you don't need an interface don't define it, if you don't need to have error package don't create it etc... so I would keep it here until needed to be used by more places